### DISCUSSION TITLE
Visualizing non-convex polygons

### DISCUSSION THREAD

**Discussion by ZiguoAtGitHub** [marcomusy/vedo#806] [Q&A] [answered]
Hi Marco,
I want to plot a H-shape face. But I got this. Maybe there is a bug. :)
![image](https://user-images.githubusercontent.com/116484589/217811903-d6335cf8-d6ac-45fe-b4b7-b01d4ff968c4.png)

"""Set custom lights to a 3D scene"""
```
from vedo import *
B,D,TF,TW,BB,TFB=500,800,50,35,650,20
plist=[(-B/2.0,D/2.0),(B/2.0,D/2.0),(B/2.0,D/2-TF),(TW/2.0,D/2-TF),(TW/2.0,-D/2+TFB),(BB/2.0,-D/2+TFB),
         (BB/2.0,-D/2),(-BB/2.0,-D/2),(-BB/2.0,-D/2+TFB),(-TW/2.0,-D/2+TFB),(-TW/2.0,D/2-TF),(-B/2.0,D/2-TF)]
ndlist=[(ip[0],ip[1],0.0) for ip in plist]
# Add light sources at the given positions
mesh=Mesh([ndlist,[(0,1,2,3,4,5,6,7,8,9,10,11,0)]])

show(mesh, __doc__, axes=1, viewup='z').close()


```

**Comment by marcomusy** [comment, marked as answer]
I would write it as:
```python
from vedo import *
B, D, TF, TW, BB, TFB = 500, 800, 50, 35, 650, 20
plist = [
    (-B / 2, D / 2),
    (B / 2, D / 2),
    (B / 2, D / 2 - TF),
    (TW / 2, D / 2 - TF),
    (TW / 2, -D / 2 + TFB),
    (BB / 2, -D / 2 + TFB),
    (BB / 2, -D / 2),
    (-BB / 2, -D / 2),
    (-BB / 2, -D / 2 + TFB),
    (-TW / 2, -D / 2 + TFB),
    (-TW / 2, D / 2 - TF),
    (-B / 2, D / 2 - TF),
]
faces = [(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11)]
msh = Mesh([plist, faces]).triangulate().c("blue4").lw(1)
show(msh, axes=1).close()
```
<img width="283" alt="Screenshot 2023-02-09 at 13 46 20" src="https://user-images.githubusercontent.com/32848391/217816760-fa07b57f-4171-4770-8641-80bcb8fdec7c.png">

do not repeat `0` to define the single face. Note that a single face which is not convex cannot be rendered, so you need `.triangulate()`.

Hope this helps.

**Reply by ZiguoAtGitHub** [reply]
Thanks for your reply. :)
I have another solution:
```
from vedo import *

B,D,TF,TW,BB,TFB=500,800,50,35,650,20
plist=[(-B/2.0,D/2.0),(B/2.0,D/2.0),(B/2.0,D/2-TF),(TW/2.0,D/2-TF),(TW/2.0,-D/2+TFB),(BB/2.0,-D/2+TFB),
         (BB/2.0,-D/2),(-BB/2.0,-D/2),(-BB/2.0,-D/2+TFB),(-TW/2.0,-D/2+TFB),(-TW/2.0,D/2-TF),(-B/2.0,D/2-TF)]
ndlist=[(ip[0],ip[1],0.0) for ip in plist]
t_coord=[((iv[0]+250)/500,(iv[1]+400)/800) for iv in ndlist]
mesh2=Mesh([ndlist,[(0,1,2,11)]]).c('blue')
mesh1=Mesh([ndlist,[(10,3,4,9)]]).c('blue')
mesh3=Mesh([ndlist,[(8,5,6,7)]]).c('blue')
mesh1.clean()
mesh2.clean()
mesh3.clean()
mesh4=merge([mesh1,mesh2,mesh3])
show(mesh4, __doc__, axes=1, viewup='y').close()

```
![image](https://user-images.githubusercontent.com/116484589/217991715-1b2e9d6a-a32d-4896-b362-e6f8918d8426.png)

**Reply by ZiguoAtGitHub** [reply]
Oh, please forget the t_coord list which is prepared for texture. :)

**Reply by marcomusy** [reply]
yes - note that in this case you don't need `clean()`

**Reply by PinkMushroom** [reply]
> I would write it as:
> 
> ```python
> from vedo import *
> B, D, TF, TW, BB, TFB = 500, 800, 50, 35, 650, 20
> plist = [
>     (-B / 2, D / 2),
>     (B / 2, D / 2),
>     (B / 2, D / 2 - TF),
>     (TW / 2, D / 2 - TF),
>     (TW / 2, -D / 2 + TFB),
>     (BB / 2, -D / 2 + TFB),
>     (BB / 2, -D / 2),
>     (-BB / 2, -D / 2),
>     (-BB / 2, -D / 2 + TFB),
>     (-TW / 2, -D / 2 + TFB),
>     (-TW / 2, D / 2 - TF),
>     (-B / 2, D / 2 - TF),
> ]
> faces = [(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11)]
> msh = Mesh([plist, faces]).triangulate().c("blue4").lw(1)
> show(msh, axes=1).close()
> ```

Hi Marco,
Is the method you recommend for all polygons when you know all the points/shape of the polygon ?

What if you need to generate hundreds of polygons ?

I am reading in CAD data, and I do know the points in the polygon.

My initial attempt was to use a point cloud + generate_delaunay2D but that does not give the right shape (which makes sense when I think about it).
Then I tried generate_mesh which works much better but you have to know the order of the points to know how to set the invert argument.

This method seems bullet proof and also more efficient than either of the point cloud methods.

Thanks!
